src/main/java/com/epam/aidial/core/upstream/WeightedRoundRobinBalancer.java

```diff
@@ -0,0 +1,112 @@
+package com.epam.aidial.core.upstream;
+
+import com.epam.aidial.core.config.Upstream;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import javax.annotation.Nullable;
+
+@Getter
+@Slf4j
+public class WeightedRoundRobinBalancer implements Comparable<WeightedRoundRobinBalancer>, LoadBalancer<UpstreamState> {
+
+    private final int tier;
+    private final List<UpstreamState> upstreams;
+    private final String deploymentName;
+
+    private int currentIndex = -1;
+    private int currentWeight;
+
+    public WeightedRoundRobinBalancer(String deploymentName, List<Upstream> upstreams) {
+        this.deploymentName = deploymentName;
+        if (upstreams == null || upstreams.isEmpty()) {
+            throw new IllegalArgumentException("Upstream list is null or empty");
+        }
+        int tier = upstreams.get(0).getTier();
+        for (Upstream upstream : upstreams) {
+            if (upstream.getTier() != tier) {
+                throw new IllegalArgumentException("Tier mismatch");
+            }
+        }
+        this.tier = tier;
+        this.upstreams = upstreams.stream()
+                .filter(upstream -> upstream.getWeight() > 0)
+                .map(upstream -> new UpstreamState(upstream, Upstream.ERROR_THRESHOLD))
+                .sorted(Comparator.reverseOrder())
+                .toList();
+        if (this.upstreams.isEmpty()) {
+            log.warn("No available upstreams for deployment %s and tier %d".formatted(deploymentName, tier));
+        }
+    }
+
+    @Nullable
+    @Override
+    public synchronized UpstreamState get() {
```

Comment:
The also doesn't work properly if weights are prime numbers. See the test below

```
public void test() {
        Upstream upstream1 = new Upstream();
        upstream1.setWeight(59);
        upstream1.setTier(1);
        upstream1.setEndpoint("1");
        Upstream upstream2 = new Upstream();
        upstream2.setWeight(53);
        upstream2.setTier(1);
        upstream2.setEndpoint("2");
        Upstream upstream3 = new Upstream();
        upstream3.setWeight(47);
        upstream3.setTier(1);
        upstream3.setEndpoint("3");
        WeightedRoundRobinBalancer balancer = new WeightedRoundRobinBalancer("app", List.of(upstream1, upstream2, upstream3));
        Map<Upstream, Integer> map = new HashMap<>();
        for (int i = 0; i < 10; i++) {
            UpstreamState upstreamState = balancer.get();
            assertNotNull(upstreamState, "failed at " + i);
            map.put(upstreamState.getUpstream(), map.getOrDefault(upstreamState.getUpstream(), 0) + 1);
        }
        for (var e : map.entrySet()) {
            System.out.println(e.getKey().getEndpoint() + " -> " + e.getValue());
        }
    }
```

The output is
```
1 -> 8
2 -> 2
```

but should be different
```
1 -> 4
2 -> 3
3 -> 3
```